src/core/device.c

```diff
@@ -703,10 +737,12 @@ static void device_update_found_one(Device *d, DeviceFound found, DeviceFound ma
                 d->found = n;
 
                 device_found_changed(d, previous, n);
-        } else
+        } else {
                 /* We aren't running yet, let's apply the new mask to the shadow variable instead, which we'll apply as
                  * soon as we catch-up with the state. */
                 d->enumerated_found = (d->enumerated_found & ~mask) | (found & mask);
+                d->enumerated_mask |= mask;
```

Comment:
> This code path is only used if `device_update_found_one()` is called while `MANAGER_IS_RUNNING()` is false.

Yes.

> The `DEVICE_FOUND_UDEV` bit in `enumerated_mask` will be set in the call chain `device_enumerate()→device_update_found_by_sysfs()→device_update_found_one()` if and only if the device was encountered by `device_enumerate()`, in which case the bit will be set in `d->enumerated_found`, too.
> 
> Tracking `enumerated_mask` separately would only make sense if the `DEVICE_FOUND_UDEV` bit could be cleared in `enumerated_found` later in a different code path. I don't see how this could happen.

Currently, that's right. But in general, that DEVICE_FOUND_UDEV flag set in `xxx_found` means the device is ready, and it is more than the fact that the device exists. Here, I'd like to make it clear if the device has udev database or not.

Reply:
> Here, I'd like to make it clear if the device has udev database or not.

But that's exactly the meaning of  `FLAG_SET(d->enumerated_found, DEVICE_FOUND_UDEV)`, or what am I missing here?

Reply:
No. If `FLAG_SET(d->enumerated_found, DEVICE_FOUND_UDEV)` is true, then `device_is_ready()` for the device is true.
If the flag is not set, there are multiple possible situations.
- the device is removed from sysfs,
- the udev database for the device has been removed,
- the device is not ready for PID1, e.g. SYSTEMD_READY=0.

Reply:
> No. If `FLAG_SET(d->enumerated_found, DEVICE_FOUND_UDEV)` is true, then `device_is_ready()` for the device is true. If the flag is not set, there are multiple possible situations.
> 
>     * the device is removed from sysfs,

This is unrelated to `d->enumerated_found`, because udevd isn't running at this point in time. The device may have disappeared from sysfs, but udev wouldn't have noticed, so the device would still be in the udev db (provided it had the `db_persist` flag before switching root — otherwise the device hadn't been enuerated anyway).

>     * the udev database for the device has been removed,
>     * the device is not ready for PID1, e.g. SYSTEMD_READY=0.

Right, these two can't be distinguished, but AFAICS they wouldn't be distinguished with `enumerated_mask`, either??